File: javascript/easy/balancedBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    return checkBalanced(root).balanced;
};

var checkBalanced = function(root) {
    if(root == null) return { depth: 0, balanced: true };
    let leftObj = checkBalanced(root.left);
    let rightObj = checkBalanced(root.right);
    let isBalanced = leftObj.balanced && rightObj.balanced && Math.abs(leftObj.depth - rightObj.depth) <= 1;
    return { depth: Math.max(leftObj.depth, rightObj.depth)+1, balanced: isBalanced };
};

File: ruby/easy/balanced_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    return check_balanced(root) != -1
end

def check_balanced(root)
    return 0 unless root
    left = check_balanced(root.left)
    right = check_balanced(root.right)
    return (left == -1 || right == -1 || (left-right).abs > 1) ? -1 : [left, right].max + 1;
end
